astropy/modeling/core.py

```diff
@@ -2218,762 +2284,1017 @@ def _join_operator(f, g):
 
-    def __getitem__(self, index):
-        index = self.__class__._normalize_index(index)
-        model = self.__class__[index]
+        Fittable parameters maintain this list and fitters modify it.
+        """
 
-        if isinstance(index, slice):
-            param_names = model.param_names
-        else:
-            param_map = self.__class__._param_map_inverse
-            param_names = tuple(param_map[index, name]
-                                for name in model.param_names)
+        # Currently the sequence of a model's parameters must be contiguous
+        # within the _parameters array (which may be a view of a larger array,
+        # for example when taking a sub-expression of a compound model), so
+        # the assumption here is reliable:
+        if self.param_names is None:
+            raise RuntimeError("Compound model parameter interface is not "
+                              "supported\n"
+                              "until the .map_parameters() method is called.")
 
-        return model._from_existing(self, param_names)
+        self._parameters_to_array()
+        start = self._param_metrics[self.param_names[0]]['slice'].start
+        stop = self._param_metrics[self.param_names[-1]]['slice'].stop
 
-    @property
-    def submodel_names(self):
-        return self.__class__.submodel_names
+        return self._parameters[start:stop]
 
-    @sharedmethod
-    def n_submodels(self):
-        return len(self.submodel_names)
+    @parameters.setter
+    def parameters(self, value):
+        """
+        Assigning to this attribute updates the parameters array rather than
+        replacing it.
+        """
+
+        if self.param_names is None:
+            raise RuntimeError("Compound model parameter interface is not "
+                               "supported\n"
+                               "until the .map_parameters() method is called.")
+
+        start = self._param_metrics[self.param_names[0]]['slice'].start
+        stop = self._param_metrics[self.param_names[-1]]['slice'].stop
+
+        try:
+            value = np.array(value).flatten()
+            self._parameters[start:stop] = value
+        except ValueError as e:
+            raise InputParameterError(
+                "Input parameter values not compatible with the model "
+                "parameters array: {0}".format(e))
+        self._array_to_parameters()
+
+    @inverse.setter
+    def inverse(self, invmodel):
+        if not isinstance(invmodel, Model):
+            raise ValueError("Attempt to assign non model to inverse")
+        self._has_inverse = True
+        self._user_inverse = invmodel
+
+    __add__ =     _model_oper('+')
+    __sub__ =     _model_oper('-')
+    __mul__ =     _model_oper('*')
+    __truediv__ = _model_oper('/')
+    __pow__ =     _model_oper('**')
+    __or__ =      _model_oper('|')
+    __and__ =     _model_oper('&')
+    ###__mod__ =     _model_oper('%')
+
+    def map_parameters(self, namestyle=None):
```

Comment:
`namestyle` is not used, what was the intended use?

Reply:
At one point it was wondered if an alternate naming scheme should be adopted. But I think that need has gone away.